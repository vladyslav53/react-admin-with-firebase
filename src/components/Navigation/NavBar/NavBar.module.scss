.navBar {
  margin-top: 5px;
}
