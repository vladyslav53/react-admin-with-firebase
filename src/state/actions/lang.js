import { createAction } from 'redux-act';

export const LANG_SET_LOCALE = createAction('LANG_SET_LOCALE');

export const langSetLocale = locale => dispatch => {
  return dispatch(LANG_SET_LOCALE({ locale }));
};
