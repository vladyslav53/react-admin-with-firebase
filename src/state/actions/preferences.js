import { createAction } from 'redux-act';

export const PREFERENCES_SET_LOCALE = createAction('PREFERENCES_SET_LOCALE');

export const preferencesSetLocale = locale => dispatch => {
  return dispatch(PREFERENCES_SET_LOCALE({ locale }));
};
