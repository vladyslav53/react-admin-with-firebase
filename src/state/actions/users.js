import { createAction } from 'redux-act';
import uuid from 'uuid/v4';
import { toastr } from 'react-redux-toastr';

import axios from 'utils/axios';
import { firebaseError } from 'utils';
import firebase from 'firebase.js';
import { checkUserData, AUTH_UPDATE_USER_DATA } from './auth';

export const USERS_FETCH_DATA_INIT = createAction('USERS_FETCH_DATA_INIT');
export const USERS_FETCH_DATA_SUCCESS = createAction(
  'USERS_FETCH_DATA_SUCCESS'
);
export const USERS_FETCH_DATA_FAIL = createAction('USERS_FETCH_DATA_FAIL');

export const USERS_DELETE_USER_INIT = createAction('USERS_DELETE_USER_INIT');
export const USERS_DELETE_USER_SUCCESS = createAction(
  'USERS_DELETE_USER_SUCCESS'
);
export const USERS_DELETE_USER_FAIL = createAction('USERS_DELETE_USER_FAIL');

export const USERS_CLEAR_DATA = createAction('USERS_CLEAR_DATA');

export const USERS_CREATE_USER_INIT = createAction('USERS_CREATE_USER_INIT');
export const USERS_CREATE_USER_SUCCESS = createAction(
  'USERS_CREATE_USER_SUCCESS'
);
export const USERS_CREATE_USER_FAIL = createAction('USERS_CREATE_USER_FAIL');

export const USERS_MODIFY_USER_INIT = createAction('USERS_MODIFY_USER_INIT');
export const USERS_MODIFY_USER_SUCCESS = createAction(
  'USERS_MODIFY_USER_SUCCESS'
);
export const USERS_MODIFY_USER_FAIL = createAction('USERS_MODIFY_USER_FAIL');

export const USERS_CLEAN_UP = createAction('USERS_CLEAN_UP');

export const fetchUsers = () => {
  return async (dispatch, getState) => {
    dispatch(checkUserData());

    dispatch(USERS_FETCH_DATA_INIT());

    const { tenant, id } = getState().auth.userData;

    let users;

    try {
      const ref = firebase.database().ref('users');

      users = (
        await ref
          .orderByChild('tenant')
          .equalTo(tenant)
          .once('value')
      ).val();
    } catch (error) {
      toastr.error('', error);
      return dispatch(USERS_FETCH_DATA_FAIL({ error }));
    }

    const usersData = users
      ? Object.entries(users).map(([key, value]) => ({
          id: key,
          ...value
        }))
      : [];

    return dispatch(
      USERS_FETCH_DATA_SUCCESS({
        users: usersData.filter(user => user.id !== id)
      })
    );
  };
};

export const deleteUser = id => {
  return async dispatch => {
    dispatch(USERS_DELETE_USER_INIT());

    const user = firebase.auth().currentUser;

    const userToken = await user.getIdToken();

    try {
      await axios(userToken).delete(`/users/${id}`);
    } catch (error) {
      toastr.error('', error);
      return dispatch(USERS_DELETE_USER_FAIL({ error }));
    }

    toastr.success('', 'The user was deleted.');
    return dispatch(USERS_DELETE_USER_SUCCESS({ id }));
  };
};

export const clearUsersData = () => {
  return dispatch => {
    dispatch(USERS_CLEAR_DATA());
  };
};

export const createUser = ({
  name,
  email,
  location,
  file,
  createdAt,
  isAdmin
}) => {
  return async (dispatch, getState) => {
    dispatch(USERS_CREATE_USER_INIT());

    const { tenant } = getState().auth.userData;

    const user = firebase.auth().currentUser;

    const userToken = await user.getIdToken();

    const body = new FormData();

    if (file) {
      const fileExtension = file.name.split('.')[1];

      const fileName = `${uuid()}.${fileExtension}`;

      body.append('logo', file, fileName);
    }

    body.append('name', name);
    body.append('location', location);
    body.append('email', email);
    body.append('password', uuid());
    body.append('tenant', tenant);
    body.append('createdAt', createdAt);
    body.append('isAdmin', isAdmin);

    axios(userToken)
      .post('/users', body)
      .then(async response => {
        const userCreated = response.data;

        const actionCodeSettings = {
          url: process.env.REACT_APP_LOGIN_PAGE_URL,
          handleCodeInApp: true
        };

        try {
          await firebase
            .auth()
            .sendSignInLinkToEmail(email, actionCodeSettings);
        } catch (error) {
          const errorMessage = firebaseError(error.response.data.error.code);

          return dispatch(
            USERS_CREATE_USER_FAIL({
              error: errorMessage
            })
          );
        }

        toastr.success('', 'User created successfully');
        return dispatch(
          USERS_CREATE_USER_SUCCESS({
            user: userCreated
          })
        );
      })
      .catch(error => {
        const errorMessage = firebaseError(error.response.data.error.code);
        toastr.error('', errorMessage);
        return dispatch(
          USERS_CREATE_USER_FAIL({
            error: errorMessage
          })
        );
      });
  };
};

export const modifyUser = ({
  name,
  location,
  isAdmin,
  file,
  createdAt,
  id,
  isEditing,
  isProfile
}) => {
  return async (dispatch, getState) => {
    dispatch(USERS_MODIFY_USER_INIT());

    const { tenant } = getState().auth.userData;

    const user = firebase.auth().currentUser;

    const userToken = await user.getIdToken();

    const body = new FormData();

    if (file) {
      const fileExtension = file.name.split('.')[1];

      const fileName = `${uuid()}.${fileExtension}`;

      body.append('logo', file, fileName);
    }

    body.append('name', name);
    body.append('location', location);
    body.append('tenant', tenant);
    body.append('createdAt', createdAt);
    body.append('isAdmin', isAdmin);

    axios(userToken)
      .patch(`/users/${id}`, body)
      .then(response => {
        const userCreated = response.data;
        const { uid } = firebase.auth().currentUser;

        if (id === uid) {
          dispatch(AUTH_UPDATE_USER_DATA({ ...userCreated }));
        }

        if (isProfile) {
          toastr.success('', 'Profile updated successfully');
        } else if (isEditing) {
          toastr.success('', 'User updated successfully');
        }

        return dispatch(
          USERS_MODIFY_USER_SUCCESS({
            user: userCreated
          })
        );
      })
      .catch(error => {
        const errorMessage = firebaseError(error.response.data.error.code);
        toastr.error('', errorMessage);
        return dispatch(
          USERS_MODIFY_USER_FAIL({
            error: errorMessage
          })
        );
      });
  };
};

export const usersCleanUp = () => dispatch => dispatch(USERS_CLEAN_UP());
