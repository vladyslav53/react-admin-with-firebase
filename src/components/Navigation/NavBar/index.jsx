import React, { useState, useCallback } from 'react';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import { langSetLocale } from 'state/actions/lang';
import { useFormatMessage } from 'hooks';
import en from 'assets/en.png';
import es from 'assets/es.png';
import { logout } from 'state/actions/auth';
import paths from 'pages/Router/paths';
import defaultLogo from 'assets/user-default-log.svg';
import Link from '../Link';

const NavBar = ({ handleMobileToggle, asideMobileActive }) => {
  const [navMobileActive, setNavMobileActive] = useState(false);

  const { userName, logoUrl } = useSelector(
    state => ({
      userName: state.auth.userData.name,
      logoUrl: state.auth.userData.logoUrl
    }),
    shallowEqual
  );

  const dispatch = useDispatch();

  const onClickLogoutHandler = () => {
    dispatch(logout());
  };

  const onMobileToggleHandler = useCallback(() => {
    setNavMobileActive(!navMobileActive);
  }, [setNavMobileActive, navMobileActive]);

  const changeToEnglishHandler = () => {
    dispatch(langSetLocale('en'));
  };

  const changeToSpanishHandler = () => {
    dispatch(langSetLocale('es'));
  };

  return (
    <nav id="navbar-main" className="navbar is-fixed-top">
      <div className="navbar-brand">
        <a
          className="navbar-item is-hidden-desktop jb-aside-mobile-toggle"
          onClick={handleMobileToggle}
        >
          <span className="icon">
            <i
              className={classNames(
                'mdi',
                'mdi-24px',
                { 'mdi-forwardburger': !asideMobileActive },
                { 'mdi-backburger': asideMobileActive }
              )}
            />
          </span>
        </a>
      </div>
      <div className="navbar-brand is-right">
        <a
          className="navbar-item is-hidden-desktop jb-navbar-menu-toggle"
          data-target="navbar-menu"
          onClick={onMobileToggleHandler}
        >
          <span className="icon">
            <i
              className={classNames(
                'mdi',
                { 'mdi-dots-vertical': !navMobileActive },
                { 'mdi-close': navMobileActive }
              )}
            />
          </span>
        </a>
      </div>
      <div
        className={classNames('navbar-menu', 'fadeIn', 'animated', 'faster', {
          'is-active': navMobileActive
        })}
        id="navbar-menu"
      >
        <div className="navbar-end">
          <div className="navbar-item has-dropdown has-dropdown-with-icons has-divider has-user-avatar is-hoverable">
            <a className="navbar-link is-arrowless">
              <div className="is-language">
                <span>{useFormatMessage('NavBar.Language', 'Language')}</span>
              </div>
              <span className="icon">
                <i className="mdi mdi-chevron-down" />
              </span>
            </a>
            <div className="navbar-dropdown">
              <a onClick={changeToEnglishHandler} className="dropdown-item">
                <img src={en} alt="English icon" />
                {useFormatMessage('NavBar.en', 'ENGLISH')}
              </a>
              <hr className="navbar-divider" />
              <a onClick={changeToSpanishHandler} className="dropdown-item">
                <img src={es} alt="Spanish icon" />
                {useFormatMessage('NavBar.es', 'SPANISH')}
              </a>
            </div>
          </div>
          <div className="navbar-item has-dropdown has-dropdown-with-icons has-divider has-user-avatar is-hoverable">
            <a className="navbar-link is-arrowless">
              <div className="is-user-avatar">
                <img src={logoUrl || defaultLogo} alt="User profile" />
              </div>
              <div className="is-user-name">
                <span>{userName}</span>
              </div>
              <span className="icon">
                <i className="mdi mdi-chevron-down" />
              </span>
            </a>
            <div className="navbar-dropdown">
              <Link to={paths.PROFILE} onClick={onMobileToggleHandler}>
                <span className="icon">
                  <i className="mdi mdi-account" />
                </span>
                <span>{useFormatMessage('NavBar.profile', 'Profile')}</span>
              </Link>
              <hr className="navbar-divider" />
              <a
                className="navbar-item"
                id="logout"
                onClick={onClickLogoutHandler}
              >
                <span className="icon">
                  <i className="mdi mdi-logout" />
                </span>
                <span>{useFormatMessage('NavBar.logOut', 'Log Out')}</span>
              </a>
            </div>
          </div>
        </div>
      </div>
    </nav>
  );
};

NavBar.propTypes = {
  handleMobileToggle: PropTypes.func.isRequired,
  asideMobileActive: PropTypes.bool
};

export default NavBar;
