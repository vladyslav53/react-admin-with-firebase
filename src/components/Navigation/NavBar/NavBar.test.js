import React from 'react';
import * as reactRedux from 'react-redux';
import { Link } from 'react-router-dom';

import * as actions from '../../../state/actions/auth';
import NavBar from '.';

const onHandleMobile = jest.fn();

describe('<NavBar /> rendering', () => {
  it('should render without crashing', () => {
    const { component } = shallowWithProvider(
      <NavBar handleMobileToggle={onHandleMobile} />
    )({
      auth: {
        userData: {}
      }
    });

    expect(component).toMatchSnapshot();
  });

  it('should render the links correctly', () => {
    const { component } = mountWithProvider(
      <NavBar handleMobileToggle={onHandleMobile} />
    )({
      auth: {
        userData: {}
      }
    });

    expect(component.find(Link)).toHaveLength(1);
  });
});

describe('<Bar /> actions', () => {
  const dispatchMock = jest.fn();

  beforeEach(() => {
    jest
      .spyOn(reactRedux, 'useDispatch')
      .mockImplementation(() => dispatchMock);
    jest.spyOn(actions, 'logout').mockImplementation(jest.fn);
  });

  it('should dispatch logout action when the user tries to logout', () => {
    const { component } = mountWithProvider(
      <NavBar handleMobileToggle={onHandleMobile} />
    )({
      auth: {
        userData: {}
      }
    });

    component.find('#logout').simulate('click');

    expect(actions.logout).toHaveBeenCalled();
  });
});
