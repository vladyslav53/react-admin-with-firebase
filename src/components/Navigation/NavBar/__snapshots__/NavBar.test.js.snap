// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavBar /> rendering should render without crashing 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <NavBar
      handleMobileToggle={[MockFunction]}
    />
  </Provider>
</Router>
`;
